Enable `backward` on `_foreach_zero_`

crcrpar opened the pull request:
Currently torchgen cannot find an appropriate `DifferentiabilityInfo` for `_foreach_zero_` because `gen_foreach_derivativeinfo` doesn't correctly make use of `functional_info_by_signature` and `differentiability_infos`, and `is_reference_for_foreach` a bit too strict to `_foreach_zero_`.

Generated code in `VariableType`
```c++
void _foreach_zero_(c10::DispatchKeySet ks, at::TensorList self) {
  auto self_ = unpack(self, "self", 0);
  [[maybe_unused]] auto _any_requires_grad = compute_requires_grad( self );

  std::vector<c10::optional<at::Tensor>> original_selfs(self.size());
  std::vector<std::shared_ptr<ZeroBackward0>> grad_fns;
  if (_any_requires_grad) {
    for (const auto& i : c10::irange( self.size() )) {
      const auto ith_requires_grad = compute_requires_grad(self[i]);
      check_inplace(self[i], ith_requires_grad);
      grad_fns.push_back([&]() -> std::shared_ptr<ZeroBackward0> {
          if (!ith_requires_grad) {
              return nullptr;
          } else {
              auto grad_fn = std::shared_ptr<ZeroBackward0>(new ZeroBackward0(), deleteNode);
              grad_fn->set_next_edges(collect_next_edges( self[i] ));
              return grad_fn;
          }
      }());
    }
  }
  #ifndef NDEBUG
  std::vector<c10::optional<Storage>> self__storage_saved(self_.size());
  for (const Tensor& tensor : self_)
    self__storage_saved.push_back(
      tensor.has_storage() ? c10::optional<Storage>(tensor.storage()) : c10::nullopt);
  std::vector<c10::intrusive_ptr<TensorImpl>> self__impl_saved(self_.size());
  for (size_t i=0; i<self_.size(); i++)
    if (self_[i].defined()) self__impl_saved[i] = self_[i].getIntrusivePtr();
  #endif
  {
    at::AutoDispatchBelowAutograd guard;
    at::redispatch::_foreach_zero_(ks & c10::after_autograd_keyset, self_);
  }
  #ifndef NDEBUG
  for (size_t i=0; i<self_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (self__storage_saved[i].has_value() && !at::impl::tensorlist_has_dispatch(self_))
      TORCH_INTERNAL_ASSERT(self__storage_saved[i].value().is_alias_of(self_[i].storage()));
  }
  for (size_t i=0; i<self_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (self__impl_saved[i] && !at::impl::tensorlist_has_dispatch(self_))
      TORCH_INTERNAL_ASSERT(self__impl_saved[i] == self_[i].getIntrusivePtr());
  }
  #endif
  if (!grad_fns.empty()) {
      auto differentiable_outputs = flatten_tensor_args( self );
      TORCH_INTERNAL_ASSERT(differentiable_outputs.size() == grad_fns.size());
      for (const auto& i : c10::irange(grad_fns.size())) {
          auto grad_fn = grad_fns[i];
          if (grad_fn != nullptr) {
              rebase_history(differentiable_outputs[i], grad_fns[i]);
          }
      }
  }
}
```


Rel:
- #58833 
- #96405

crcrpar committed:
optionally consult `non_functional_info_by_signature`

as especially `_foreach_zero_`'s reference `DifferentiabilityInfo` seems
to be only available in it, not `functional_info_by_signature`.

Also, made `gen_foreach_derivativeinfo` a bit cleaner as passing both
`differentiability_infors` and `functional_info_by_signature` wouldn't
make less sense than passing `functional_info_by_signature` and
`non_functional_info_by_signature`.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
add _foreach_zero_ test

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented on torchgen/api/autograd.py:
```diff
@@ -316,53 +316,53 @@ def is_reference_for_foreach(
     functional_info_by_signature: Dict[
         FunctionSchema, Dict[str, DifferentiabilityInfo]
     ],
+    non_functional_info_by_signature: Dict[
```
ah this is technically breaking backward compatibility but this function from my perspective isn't general enough for wide adoption.

Would you be okay with suppressing the linter https://github.com/pytorch/test-infra/wiki/BC-Linter#suppression?

soulitzer replied:
suppressing sounds ok!

soulitzer commented on torchgen/api/autograd.py:
```diff
@@ -316,53 +316,53 @@ def is_reference_for_foreach(
-            )
+    return f.func.name.name.base.split("_foreach_")[
+        -1
+    ] == function_schema.name.name.base and all(
```
haha this is an interesting line break, took me a second to parse this
any way we could keep it on the same line as before?

soulitzer replied:
Also we should probably ignore the `and not function_schema.name.name.inplace` condition only  in the case of zero_?

We could have a single set that is used both here and below in `gen_foreach_derivativeinfo`.

crcrpar committed:
remove print from test

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

soulitzer commented on test/test_foreach.py:
```diff
@@ -900,6 +900,31 @@ def test_outplace_with_invalid_grads(self, device, dtype, op):
         self.assertIsNotNone(sample.input[0].grad)
         self.assertIsNone(sample.input[1].grad)
 
+    # note(crcrpar): this is clumsy but we don't have out-place `_foreach_zero` so make it a special case.
```
Could we make an OpInfo for the inplace version? (we have an OpInfo for the inplace `zero_`)

crcrpar replied:
Yes we could but I thought this would be simpler (sorry for being lazy haha)

Some changes would be needed in `ForeachFuncInfo`. Currently ForeachFuncInfo I wrote previously assumes that out-place is always exists while in-place not. This restriction derives from around
https://github.com/pytorch/pytorch/blob/49c8a0cad0dc314fba05c0a7d909df0b2ab991ab/torch/testing/_internal/opinfo/core.py#L2634-L2644

soulitzer replied:
Hmm I'm still concerned about how this makes it harder to test things consistently across the the ops.

I'd be okay with a special case in `get_foreach_method_names`, e.g:
```
if name == "zero_":
    <do something else>
```

soulitzer reviewed (commented):
Thanks for the fix!

crcrpar commented:
@pytorchbot label suppress-bc-linter

crcrpar committed:
stricter `is_reference_for_foreach` func

by creating allowlist of foreach functions whose reference function is
somehow inplace

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
ForeachFuncInfo("zero")

with a fake `op` as arg to OpInfo dunder init

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

soulitzer reviewed (approved):
OK!

crcrpar commented:
The failure refers to https://github.com/pytorch/pytorch/issues/100209🤔

crcrpar commented:
@pytorchbot merge -r
